### PR TITLE
bpf: Reset CT entry for CT_REOPENED

### LINKED ISSUE #27762: CI: Conformance E2E: client-egress-l7-named-port/pod-to-pod: command terminated with exit code 28 (timeout)
## CI failure

Matrix entry: (5, 5.15-20230420.212204, iptables, true, disabled, dsr, true, true, true)

```
ℹ️  📜 Applying CiliumNetworkPolicy 'client-egress-only-dns' to namespace 'cilium-test'..
  ℹ️  📜 Applying CiliumNetworkPolicy 'client-egress-l7-http-named-port' to namespace 'cilium-test'..
  [-] Scenario [client-egress-l7-named-port/pod-to-pod]
  [.] Action [client-egress-l7-named-port/pod-to-pod/curl-ipv4-0: cilium-test/client-6b4b857d98-hnf49 (10.244.1.76) -> cilium-test/echo-other-node-8f5d547bf-sgh62 (10.244.2.69:8080)]
  [.] Action [client-egress-l7-named-port/pod-to-pod/curl-ipv6-0: cilium-test/client-6b4b857d98-hnf49 (fd00:10:244:1::47c6) -> cilium-test/echo-other-node-8f5d547bf-sgh62 (fd00:10:244:2::294e:8080)]
  [.] Action [client-egress-l7-named-port/pod-to-pod/curl-ipv4-1: cilium-test/client-6b4b857d98-hnf49 (10.244.1.76) -> cilium-test/echo-same-node-7c978647dc-xr2kt (10.244.1.111:8080)]
  [.] Action [client-egress-l7-named-port/pod-to-pod/curl-ipv6-1: cilium-test/client-6b4b857d98-hnf49 (fd00:10:244:1::47c6) -> cilium-test/echo-same-node-7c978647dc-xr2kt (fd00:10:244:1::19cd:8080)]
  [.] Action [client-egress-l7-named-port/pod-to-pod/curl-ipv4-2: cilium-test/client2-646b88fb9b-x8j2f (10.244.1.44) -> cilium-test/echo-other-node-8f5d547bf-sgh62 (10.244.2.69:8080)]
  ❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 http://10.244.2.69:8080/" failed: command terminated with exit code 28
  ℹ️  curl output:


📋 Test Report
connectivity test failed: 1 tests failed
❌ 1/53 tests failed (1/575 actions), 8 tests skipped, 0 scenarios skipped:
Test [client-egress-l7-named-port]:
  ❌ client-egress-l7-named-port/pod-to-pod/curl-ipv4-2: cilium-test/client2-646b88fb9b-x8j2f (10.244.1.44) -> cilium-test/echo-other-node-8f5d547bf-sgh62 (10.244.2.69:8080)
Error: Process completed with exit code 1.
```

Looks similar to https://github.com/cilium/cilium/issues/27672 but not related to multicluster.

From a quick sysdump check, there seems to be something wrong with Envoy and the http conneciton:

```
$ cat hubble-flows-cilium-* | hubble observe --input-file - --numeric --to-ip 10.244.2.69 --port 8080 --from-ip 10.244.1.44 --type l7
Aug 28 17:16:59.869: 10.244.1.44:44528 (ID:11576) -> 10.244.2.69:8080 (ID:25450) http-request FORWARDED (HTTP/1.1 GET http://10.244.2.69:8080/)
Aug 28 17:17:21.203: 10.244.1.44:54076 (ID:11576) -> 10.244.2.69:8080 (ID:25450) http-request FORWARDED (HTTP/1.1 GET http://10.244.2.69:8080/)
Aug 28 17:15:50.297: 10.244.1.44:37878 (ID:11576) -> 10.244.2.69:8080 (ID:25450) http-request FORWARDED (HTTP/1.1 GET http://10.244.2.69:8080/public)
Aug 28 17:15:50.377: 10.244.1.44:45332 (ID:11576) -> 10.244.2.69:8080 (ID:25450) http-request DROPPED (HTTP/1.1 GET http://10.244.2.69:8080/private)
Aug 28 17:15:50.453: 10.244.1.44:45348 (ID:11576) -> 10.244.2.69:8080 (ID:25450) http-request FORWARDED (HTTP/1.1 GET http://10.244.2.69:8080/private)
Aug 28 17:16:18.302: 10.244.1.44:45148 (ID:11576) -> 10.244.2.69:8080 (ID:25450) http-request FORWARDED (HTTP/1.1 GET http://10.244.2.69:8080/public)
Aug 28 17:16:18.375: 10.244.1.44:45162 (ID:11576) -> 10.244.2.69:8080 (ID:25450) http-request DROPPED (HTTP/1.1 GET http://10.244.2.69:8080/private)
Aug 28 17:16:18.445: 10.244.1.44:45166 (ID:11576) -> 10.244.2.69:8080 (ID:25450) http-request FORWARDED (HTTP/1.1 GET http://10.244.2.69:8080/private)
```

[cilium-sysdump-5-final.zip](https://github.com/cilium/cilium/files/12458796/cilium-sysdump-5-final.zip)

### PR DESCRIPTION
**Pull request by jrajahalme**
Reset the CT entry when an old entry has been found when looking up an entry for TCP SYN packet (CT_REOPENED). This way the old entry has no effect to the entry for the new connection and security identity, as well as proxy and nodeport related flags get set correctly for the new connection.

This prevents failing proxy redirect when the CT entry of a non-redirected connection would have been reused for a redirected connection, which has happened in the CI where an egress L7 policy was applied and an old CT entry from time without the L7 policy existed.

For this bug to trigger a pod needs to open a new TCP connection using the same (ephemeral) source port to the same destination before and after a change in policy adding an L7 policy applicable to that connection. While this is a bug that has surfaced in CI, this bug is less likely to be hit when policies are more stable. Given this we need further discussion about backporting this bug fix to the stable releases.

Fixes: #27762 (see https://github.com/cilium/cilium/issues/27762#issuecomment-2118311808)

```release-note
Datapath conntrack entries for reopened connections are fully reinitialized to fix rare L7 proxy redirect failures.
```

### CODE REVIEW

**Comment by jrajahalme** [comment]
Simplified bpf changes a bit.

**Comment by jrajahalme** [comment]
bpf lint fixes.

**Comment by julianwiedmann** [comment]
Nice catch! For clarity - this change would only help TCP connections, correct? I think we need to find a solution for UDP / SCTP as well (or `CT_ESTABLISHED` in general). As something easy to backport, we could just follow the current approach and add the additional constraints [here](https://github.com/cilium/cilium/blob/db5fcdc567a4b2689a5b5f9e3873d7912d0916f7/bpf/bpf_lxc.c#L994).

But long-term I think the CT lookups from `bpf_lxc` should follow the approach we've already started for [Service-related CT entries](https://github.com/cilium/cilium/blob/db5fcdc567a4b2689a5b5f9e3873d7912d0916f7/bpf/lib/conntrack.h#L248), and use more precise matching in the low-level conntrack code. Which is also why I would prefer if we split the `nodeport.h` changes off into a separate patch (or even PR?). There's a good chance that the lookup code there is already good enough ...

**Comment by jrajahalme** [comment]
> Nice catch! For clarity - this change would only help TCP connections, correct? I think we need to find a solution for UDP / SCTP as well (or `CT_ESTABLISHED` in general). As something easy to backport, we could just follow the current approach and add the additional constraints [here](https://github.com/cilium/cilium/blob/db5fcdc567a4b2689a5b5f9e3873d7912d0916f7/bpf/bpf_lxc.c#L994).
> 

I considered this approach, but concluded that it would easily result into more brittle code, and this simpler approach is better, for the following reasons:
- we'd have to expand the comparison logic to cover all the possible differences between the old entry and the new state
- this could be extracted to a helper function
- but still the helper function could get stale
- `CT_REOPENED` happens only for new connections, and we already typically create new entries for new connections; it is simpler to just do the `CT_NEW` behavior for `CT_REOPENED`

Given this it might be better to get rid of `CT_REOPENED` altogether and only return `CT_NEW` though?

Current `lib/conntrack.h` updates the given `ct_state` for `CT_REOPENED`, so we have to explicitly clear it. would be simpler if `lib/conntrack.h` simply returned `CT_NEW` (without touching the given `*ct_state`), if the policy always is to reset the CT entry anyway?

As for changes for CT_ESTABLISHED: changes in proxy redirection mid-connection will not work, but currently policy changes break existing connections for which redirection status changes, due to forward direction immediately behaving accoring to the new policy, while return direction takes the redirection status from the CT entry, that is not currently updated for `CT_ESTABLISHED`. I guess this breaks DNS proxy for UDP across policy changes, and updating a stale CT entry for `CT_ESTABLISHED` would fix this.

> But long-term I think the CT lookups from `bpf_lxc` should follow the approach we've already started for [Service-related CT entries](https://github.com/cilium/cilium/blob/db5fcdc567a4b2689a5b5f9e3873d7912d0916f7/bpf/lib/conntrack.h#L248), and use more precise matching in the low-level conntrack code. 

Looks like `ct_entry_matches_types()` would have to to get much more complex if proxy redirection would be added in there?

File: bpf/bpf_lxc.c [hunk 1]
```diff
@@ -1979,28 +1972,20 @@ ipv4_policy(struct __ctx_buff *ctx, struct iphdr *ip4, int ifindex, __u32 src_la
...
-#ifdef ENABLE_NODEPORT
+	/* Reset the entry also for CT_REOPENED. This way any difference in source security id,
+	 * proxy redirect, dsr, or nodeport gets set correctly.
+	 */
 	if (ret == CT_NEW || ret == CT_REOPENED) {
-# ifdef ENABLE_DSR
-		/* Clear .dsr_internal flag for old connections: */
-		if (ret == CT_REOPENED && ct_state->dsr_internal)
-			ct_update_dsr(get_ct_map4(tuple), tuple, false);
-# endif /* ENABLE_DSR */
```
Comment (julianwiedmann) [inline bpf/bpf_lxc.c:1988]: Let's address this part separately (https://github.com/cilium/cilium/pull/32642), there's more to it.

File: bpf/lib/nodeport.h [hunk 2]
```diff
@@ -2332,9 +2334,6 @@ nodeport_dsr_ingress_ipv4(struct __ctx_buff *ctx, struct ipv4_ct_tuple *tuple,
 			      CT_ENTRY_DSR, NULL, &monitor);
 	switch (ret) {
 	case CT_NEW:
-	/* Maybe we can be a bit more selective about CT_REOPENED?
-	 * But we have to assume that both the CT and the SNAT entry are stale.
-	 */
 	case CT_REOPENED:
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h:2337]: Removing the comment without any explanation why it no longer applies, doesn't feel very helpful.

File: bpf/lib/nodeport.h [hunk 3]
```diff
@@ -2345,9 +2344,8 @@ nodeport_dsr_ingress_ipv4(struct __ctx_buff *ctx, struct ipv4_ct_tuple *tuple,
 			return DROP_INVALID;
 
 		ct_state_new.src_sec_id = WORLD_IPV4_ID;
-		ct_state_new.dsr_internal = 1;
-		ct_state_new.proxy_redirect = 0;
-		ct_state_new.from_l7lb = 0;
+		ct_state_new.dsr_internal = true;
+		/* all other flags are already false */
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h:2349]: This is just a cleanup, correct?

File: bpf/lib/nodeport.h [hunk 4]
```diff
@@ -2898,9 +2896,13 @@ static __always_inline int nodeport_svc_lb4(struct __ctx_buff *ctx,
...
 		switch (ret) {
+		case CT_REOPENED:
+			memset(&ct_state, 0, sizeof(ct_state));
+			ct_state.rev_nat_index = ct_state_svc.rev_nat_index;
+			fallthrough;
 		case CT_NEW:
 			ct_state.src_sec_id = src_sec_identity;
-			ct_state.node_port = 1;
+			ct_state.node_port = true;
 #ifndef HAVE_FIB_IFINDEX
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h:2906]: I still believe that teaching `ct_entry_matches_types()` about `src_sec_id` and `ifindex` for the case of `CT_ENTRY_NODEPORT` would get us much further here. And address any concerns for `CT_ESTABLISHED`.

We could literally just copy the code that fills the `ct_state` in front of the CT lookup, and say "this is what I'm expecting".

**Review by julianwiedmann — COMMENTED** [review]
ty! I took a closer look at the nodeport.h changes.

**Comment by julianwiedmann** [comment]
> Given this it might be better to get rid of `CT_REOPENED` altogether and only return `CT_NEW` though?
> 
> Current `lib/conntrack.h` updates the given `ct_state` for `CT_REOPENED`, so we have to explicitly clear it. would be simpler if `lib/conntrack.h` simply returned `CT_NEW` (without touching the given `*ct_state`), if the policy always is to reset the CT entry anyway?
> 

If we follow your change suggestion here, that would be my conclusion as well. We're already [clearing](https://github.com/cilium/cilium/blob/598209ebb8135b591e0a73d4422f15edc4b2789d/bpf/lib/conntrack.h#L318) quite a few parts of the entry, so why not go all the way. But I don't have sufficient historical context on why we've been introduced `CT_REOPENED` in the first place. Therefore I'm a bit hesitant to just drop it when it feels inconvenient.

> As for changes for CT_ESTABLISHED: changes in proxy redirection mid-connection will not work, but currently policy changes break existing connections for which redirection status changes, due to forward direction immediately behaving accoring to the new policy, while return direction takes the redirection status from the CT entry, that is not currently updated for `CT_ESTABLISHED`. I guess this breaks DNS proxy for UDP across policy changes as well, and updating a stale CT entry for `CT_ESTABLISHED` would fix this.

Ack. So we need to decide whether this PR is only about fixing the L7 proxy aspect, or wants to address fuzzy CT lookups as a whole.

File: bpf/bpf_lxc.c (hunk 1 above)
Comment (jrajahalme) [reply bpf/bpf_lxc.c:1988]: This was conditionally setting the flag to false, removed this as now the flag is false by default. IMO removal of this is the right thing to do in this PR (no functional change for the parts that were already updated for `CT_REOPENED`.

File: bpf/lib/nodeport.h (hunk 2 above)
Comment (jrajahalme) [reply bpf/lib/nodeport.h:2337]: This PR changes to reset the CT entry for `CT_REOPENED`, so there is no longer any concern for being more selective?

File: bpf/lib/nodeport.h (hunk 3 above)
Comment (jrajahalme) [reply bpf/lib/nodeport.h:2349]: Yes, stemming from the fact that other sites were missing these explicit `= 0` statements.

File: bpf/lib/nodeport.h (hunk 4 above)
Comment (jrajahalme) [reply bpf/lib/nodeport.h:2906]: So you mean the `ct_*` function would insert the new entry then?

**Comment by jrajahalme** [comment]
> 
> Ack. So we need to decide whether this PR is only about fixing the L7 proxy aspect, or wants to address fuzzy CT lookups as a whole.

I would like this PR to be a backportable bug fix for the l7 proxy redirect issue/flake. In that light maybe I should remove the cleanups as well to make this truly minimal?

THB, I don't know if the nodeport.h changes are absolutely necessary for the l7 proxy redirect fix, though.

File: bpf/lib/nodeport.h (hunk 4 above)
Comment (jrajahalme) [reply bpf/lib/nodeport.h:2906]: We can't do policy lookup before CT lookup, as we can do policy lookup on a forward direction only, and a CT lookup is needed to know if a packet is in forward or reply direction. Also, some of the CT entry contents come from the policy lookup, so it does not seem feasible to be able to give a full expected CT entry before the CT lookup.

**Comment by jrajahalme** [comment]
> > Given this it might be better to get rid of `CT_REOPENED` altogether and only return `CT_NEW` though?
> > Current `lib/conntrack.h` updates the given `ct_state` for `CT_REOPENED`, so we have to explicitly clear it. would be simpler if `lib/conntrack.h` simply returned `CT_NEW` (without touching the given `*ct_state`), if the policy always is to reset the CT entry anyway?
> 
> If we follow your change suggestion here, that would be my conclusion as well. We're already [clearing](https://github.com/cilium/cilium/blob/598209ebb8135b591e0a73d4422f15edc4b2789d/bpf/lib/conntrack.h#L318) quite a few parts of the entry, so why not go all the way. But I don't have sufficient historical context on why we've introduced `CT_REOPENED` in the first place. Therefore I'm a bit hesitant to just drop it when it feels inconvenient.
> 

Here's the commit that added CT_REOPENED` (https://github.com/cilium/cilium/pull/13340):
```
commit ada53a91a98618d24dad2abda97dd9c60923003c
Author: Zang Li <zangli@google.com>
Date:   Mon Sep 28 12:15:14 2020 -0700

    Add the return code CT_REOPENED for conntrack table lookup.
    
    When a tcp SYN packet hits a conntrack entry that is in close-wait state,
    the entry timeout will be modified and the entry will be reopened. But
    because the return code is CT_ESTABLISHED, policy-verdict event will not
    be generated even this is a new connection. The patch fix the issue by
    returning the code CT_REOPENED instead in this case. Except for policy
    verdict events, all other handlings are the same as if CT_ESTABLISHED
    is returned.
```
It looks to me `CT_REOPENED` was added to get policy verdicts also for re-opened connections. Furthermore, from the discussion therein I'd conclude that `CT_REOPENED` is in fact more likely to be a new connection (due to CT entry lingering longer than TCP state at the endpoints) rather than the same connection really being re-opened from TCP viewpoint. @joestringer argued in the PR that even a real TCP re-open should likely be policy enforced like a new connection. This is doubly so if there is a policy change between the original `SYN` and the new one!

**Comment by jrajahalme** [comment]
Superseded by https://github.com/cilium/cilium/pull/32653

**Pull request closed without merging**